Added Apache Pulsar Processors and Controller Service

david-streamlio opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

david-streamlio commented:
One single squashed commit, with both NIFI-4908 & NIFI-4914 changes in it.

MikeThomsen commented:
> Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.20.1:test (default-test) on project nifi-pulsar-processors: There are test failures.

Related to how you set up the expression language support. The boolean version of the method call is deprecated and replaced with one that uses an enum in 1.7.

david-streamlio commented:
So, is this a change in the 1.7.x code base or is it already in the 1.6.0 code?  I created my fork back on Feb 22nd based on the 1.6.0-SNAPSHOT branch, which does not have these enums.  Should I create a new fork of 1.6.x?  of 1.7.x?  Please advise

MikeThomsen commented:
I think it made it into "1.6.0-SNAPSHOT" after RC3 was cut. You have a bunch of errors in there, but they look like this:

> [ERROR] testSingleRecordSuccess(org.apache.nifi.processors.pulsar.pubsub.async.TestAsyncPublishPulsarRecord_1_X)  Time elapsed: 0.011 s  <<< FAILURE!
java.lang.AssertionError: java.lang.IllegalStateException: Attempting to evaluate expression language for topic using flow file attributes but the scope evaluation is set to NONE. The proper scope should be set in the property descriptor using PropertyDescriptor.Builder.expressionLanguageSupported(ExpressionLanguageScope)
	at org.apache.nifi.processors.pulsar.pubsub.async.TestAsyncPublishPulsarRecord_1_X.testSingleRecordSuccess(TestAsyncPublishPulsarRecord_1_X.java:87)
Caused by: java.lang.IllegalStateException: Attempting to evaluate expression language for topic using flow file attributes but the scope evaluation is set to NONE. The proper scope should be set in the property descriptor using PropertyDescriptor.Builder.expressionLanguageSupported(ExpressionLanguageScope)

david-streamlio commented:
I don't see an upstream branch labeled 1.6.0....which upstream branch should I merge into my local branch?

MikeThomsen commented:
@david-streamlio you brought in a whole lot of commits from other people with that merge from upstream/master. I checked out your branch and did a rebase on it (`git rebase master`) and that seemed to clear it up. I would recommend doing that on your branch locally and verifying that that "merge commit" with the commit message `Merge remote-tracking branch 'upstream/master' into NIFI-4914` goes away.

As a rule of thumb, this is how you want to do this sort of update:

1. git checkout master
2. git pull upstream master
3. git checkout YOUR_BRANCH
4. git rebase master

Once you've done that, the last command will replay your commits on top of the most recent version of upstream/master.

joewitt commented:
@MikeThomsen just wanted to clarify the pointers here..  Did we break an API that existing prior to 1.6 release that we should not have or was this a newer thing being leveraged that just happened to not settle until the release itself and thus requires a little rework in the PR.  Just want to make sure what we're putting @david-streamlio through here (and thanks for sticking with it david) is isolated.

MikeThomsen commented:
@joewitt something changed with the way expression language support is declared in the `PropertyDescriptor.Builder`. TL;DR when you use the deprecated `expressionLanguageSupported(boolean)` method, it doesn't seem to set up the new internal structure correctly.

pvillard31 commented:
To provide more details here: after the release of 1.6.0 (with #2205), we deprecated the use of ``expressionLanguageSupported(boolean)`` in favor of ``expressionLanguageSupported(ExpressionLanguageScope)``. Using the deprecated method does not break anything in the code itself, the only difference is just in the test framework where we are more strict checking if the appropriate scope has been set on the property description. If not, the unit test will fail.

Most of the existing opened PRs introducing new properties that support expression language without using the new scope might introduce unit test failures. The only thing required is to update the property with the proper scope.

Hope this clarifies things a bit.

pvillard31 commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service-nar/pom.xml:
```diff
@@ -0,0 +1,35 @@
+  	<dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-pulsar-client-service-api</artifactId>
+            <version>1.6.0-SNAPSHOT</version>
```
Needs to be updated to 1.7.0-SNAPSHOT

pvillard31 commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/pom.xml:
```diff
@@ -0,0 +1,67 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-pulsar-client-service-api</artifactId>
+            <version>1.6.0-SNAPSHOT</version>
```
Needs to be updated to 1.7.0-SNAPSHOT

pvillard31 commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/pom.xml:
```diff
@@ -0,0 +1,67 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-mock</artifactId>
+            <version>1.6.0-SNAPSHOT</version>
```
Needs to be updated to 1.7.0-SNAPSHOT

pvillard31 commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-nar/pom.xml:
```diff
@@ -0,0 +1,50 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-pulsar-client-service-api-nar</artifactId>
+            <version>1.6.0-SNAPSHOT</version>
```
Needs to be updated to 1.7.0-SNAPSHOT
and indentation to fix

pvillard31 commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-nar/pom.xml:
```diff
@@ -0,0 +1,50 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-pulsar-processors</artifactId>
+            <version>1.6.0-SNAPSHOT</version>
```
Needs to be updated to 1.7.0-SNAPSHOT

pvillard31 commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-nar/pom.xml:
```diff
@@ -0,0 +1,50 @@
+		<dependency>
+			<groupId>org.apache.nifi</groupId>
+			<artifactId>nifi-pulsar-client-service</artifactId>
+			<version>1.6.0-SNAPSHOT</version>
```
Needs to be updated to 1.7.0-SNAPSHOT
and indentation to fix

joewitt commented:
@pvillard31 thanks that makes a lot more sense to me now.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord_1_X.java:
```diff
@@ -0,0 +1,351 @@
+
+        // Clean-up and transfer session
+        try {
+            if (writer != null)
```
Needs curly brackets.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord_1_X.java:
```diff
@@ -0,0 +1,351 @@
+            if (writer != null)
+               writer.finishRecordSet();
+
+            if (rawOut != null)
```
Needs curly brackets.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord_1_X.java:
```diff
@@ -0,0 +1,351 @@
+            do {
+                done = consumerService.poll(queryTimeout, TimeUnit.SECONDS);
+
+                if (done == null)
```
Needs curly brackets.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord_1_X.java:
```diff
@@ -0,0 +1,351 @@
+
+        if (flowFile != null) {
+           session.putAttribute(flowFile, MSG_COUNT, messagesReceived.toString());
+           session.transfer(flowFile, REL_SUCCESS);
```
Missing provenance reporting.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsar_1_X.java:
```diff
@@ -0,0 +1,206 @@
+
+                   session.getProvenanceReporter().receive(flowFile, "From " + getWrappedConsumer(context).getTransitURL());
+                   session.transfer(flowFile, REL_SUCCESS);
+                   session.commit();
```
You might want to consider doing this every X number of flowfiles or based on a timer.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsar_1_X.java:
```diff
@@ -0,0 +1,206 @@
+    @OnStopped
+    public void close(final ProcessContext context) {
+
+        getLogger().info("Disconnecting Pulsar Consumer");
```
I think this would be better off as:

```
if (getLogger().isDebugLoggerEnabled()) {
    getLogger().debug(...
}
```

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsar_1_X.java:
```diff
@@ -0,0 +1,206 @@
+
+                session.getProvenanceReporter().receive(flowFile, "From " + context.getProperty(TOPIC).getValue());
+                session.transfer(flowFile, REL_SUCCESS);
+                logger.info("Created {} from {} messages received from Pulsar Server and transferred to 'success'",
```
Let's do this as debug and wrap with an if statement because this will get called a lot.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsar_1_X.java:
```diff
@@ -0,0 +1,206 @@
+                 * the data from the external source. In general, though, potential data duplication
+                 * is preferred over potential data loss.
+                 */
+                getLogger().info("Acknowledging message " + msg.getMessageId());
```
Needs to be wrapped as well.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsar_1_X.java:
```diff
@@ -0,0 +1,206 @@
+
+            } else {
+                // We didn't consume any data, so
+                session.commit();
```
You don't need to worry about this, AbstractProcessor will do it for you.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/InFlightMessageMonitor.java:
```diff
@@ -0,0 +1,67 @@
+ * Helper class to monitor the asynchronous submission of a large number
+ * of records to Apache Pulsar.
+ *
+ * @author david
```
Please remove.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord_1_X.java:
```diff
@@ -0,0 +1,294 @@
+        }
+
+        // Read the contents of the FlowFile into a byte array
+        final byte[] messageContent = new byte[(int) flowFile.getSize()];
```
This could break in some use cases. For example, I have processed 50GB text files using NiFi before. Not a common use case, but I'd like to know what others think here.

@bbende @markap14 @alopresto @pvillard31 @mattyb149 can one of you take a look at this and weigh in on whether there is a better way to handle the input here to accommodate very large record sets?

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord_1_X.java:
```diff
@@ -0,0 +1,294 @@
+        // Nothing to do, so skip this Flow file.
+        if (messageContent == null || messageContent.length < 1) {
+            session.transfer(flowFile, REL_SUCCESS);
+            return;
```
I would put `msg.count = 0` on this before sending it out so that people checking through the provenance history can find that this flowfile had no records in it.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord_1_X.java:
```diff
@@ -0,0 +1,294 @@
+               handleAsync(bundle, session, flowFile, topic);
+           } else {
+               messagesSent.addAndGet(send(producer, writerFactory, schema, recordSet));
+               session.putAttribute(flowFile, MSG_COUNT, messagesSent.get() + "");
```
These should be `flowfile = session.putAttribute(...` Otherwise AFAIK it's going to drop the new attribute.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord_1_X.java:
```diff
@@ -0,0 +1,294 @@
+     */
+    protected void sendAsync(Producer producer, ProcessSession session, FlowFile flowFile, InFlightMessageMonitor monitor) {
+
+        if (monitor == null || monitor.getRecords().isEmpty())
```
Should have curly brackets.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord_1_X.java:
```diff
@@ -0,0 +1,294 @@
+           monitor.getLatch().await();
+
+           if (monitor.getSuccessCounter().intValue() > 0) {
+               session.putAttribute(flowFile, MSG_COUNT, monitor.getSuccessCounter().get() + "");
```
`flowFile = session.putAttribute`

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord_1_X.java:
```diff
@@ -0,0 +1,294 @@
+              }
+
+              failureFlowFile = session.write(failureFlowFile, out -> out.write(sb.toString().trim().getBytes()));
+              session.putAttribute(failureFlowFile, MSG_COUNT, monitor.getFailureCounter().get() + "");
```
`String.valueOf(monitor.getFailureCounter().get())` would better.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsar_1_X.java:
```diff
@@ -0,0 +1,151 @@
+        final byte[] messageContent = baos.toByteArray();
+        // Nothing to do, so skip this Flow file.
+        if (messageContent == null || messageContent.length < 1) {
+            session.transfer(flowFile, REL_SUCCESS);
```
I think you should have `msg.count = 0` here.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/RecordBasedConst.java:
```diff
@@ -0,0 +1,46 @@
+            .required(true)
+            .build();
+
+    private RecordBasedConst() {
```
Not needed.

MikeThomsen commented on nifi-nar-bundles/pom.xml:
```diff
@@ -93,6 +93,7 @@
         <module>nifi-spark-bundle</module>
         <module>nifi-atlas-bundle</module>
         <module>nifi-druid-bundle</module>
+        <module>nifi-pulsar-bundle</module>
```
You need to add a dependency declaration for the NAR in `nifi-assembly/pom.xml`. Just putting that note here since there's no really good place to track that in the review.

MikeThomsen commented on nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/pom.xml:
```diff
@@ -0,0 +1,67 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-pulsar-client-service-api</artifactId>
+            <version>1.7.0-SNAPSHOT</version>
```
This should be marked as provided as well.

MikeThomsen commented:
When I added it to nifi-assembly, I got this error on startup:

```
java.util.ServiceConfigurationError: org.apache.nifi.processor.Processor: Provider org.apache.nifi.processors.pulsar.pubsub.ConsumePulsarRecord_1_X could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:232)
	at java.util.ServiceLoader.access$100(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:384)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.apache.nifi.nar.ExtensionManager.loadExtensions(ExtensionManager.java:148)
	at org.apache.nifi.nar.ExtensionManager.discoverExtensions(ExtensionManager.java:123)
	at org.apache.nifi.web.server.JettyServer.start(JettyServer.java:771)
	at org.apache.nifi.NiFi.<init>(NiFi.java:157)
	at org.apache.nifi.NiFi.<init>(NiFi.java:71)
	at org.apache.nifi.NiFi.main(NiFi.java:292)
Caused by: java.lang.NoClassDefFoundError: org/apache/nifi/serialization/MalformedRecordException
	at java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
	at java.lang.Class.getConstructor0(Class.java:3075)
	at java.lang.Class.newInstance(Class.java:412)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380)
	... 8 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.apache.nifi.serialization.MalformedRecordException
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 13 common frames omitted
```

Your declaration for `nifi-record` looks right, so we'll have to dig deeper once you get these changes done.

MikeThomsen commented:
@david-streamlio Haven't heard from you in almost 3 weeks. Have you had a chance to work on any of the code review changes?

david-streamlio commented:
@MikeThomsen I will work on those changes now and have them completed ASAP....Thanks for your patience

MikeThomsen commented:
@david-streamlio don't feel rushed, I was just checking in to see if you were still tracking this.

mattyb149 commented:
Something is a bit wonky with this latest push, perhaps a bad merge?

david-streamlio commented:
Yea, I'm not sure what happened

First I did the following 4 steps you recommended:

As a rule of thumb, this is how you want to do this sort of update:

git checkout master
git pull upstream master
git checkout YOUR_BRANCH
git rebase master

Then I checkout made the changes, ran mvn -Pcontrib-check clean install and had a clean install. When I tried to do a git push, I got the following error

Davids-MacBook-Pro:nifi david$ git push origin
To https://github.com/david-streamlio/nifi.git
 ! [rejected]            NIFI-4914 -> NIFI-4914 (non-fast-forward)
error: failed to push some refs to 'https://github.com/david-streamlio/nifi.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.


So, I fixed the conflicts, committed them and pushed again.

mattyb149 commented:
Probably needed a git rebase --continue instead...

MikeThomsen commented:
Yeah, I think @mattyb149 is right about that. I just pulled your branch again, did a rebase against master and there was an explosion of conflicts. Once you resolve those, you have to do `git rebase --continue` to finish the rebasing.

david-streamlio commented:
My latest commit passed all the tests, and is showing no conflicts with the base branch. But you are seeing "an explosion of conflicts".  So I am confused....What steps do you need me to perform in order to get this PR in an acceptable state?

MikeThomsen commented:
I pulled your branch and got a lot of merge conflicts when I tried rebasing against master. I am pretty sure that @mattyb149 is right or on the right track. When you followed those four steps, did you adjust the name `upstream` to match whatever you call the remote that points to `https://github.com/apache/nifi`? Because if you did all of that, you would have had a real fun time with all of the merge conflicts on rebasing :)

MikeThomsen commented:
Also, run `git rebase --continue` on that branch and tell us what happens. Copy and paste the output as "code" in the combox (code Markdown helper is the <> button)

david-streamlio commented:
run the rebase against my 'master' branch or my 'NIFI-4914' branch?

david-streamlio commented:
`Davids-MacBook-Pro:nifi david$ git checkout NIFI-4914
Switched to branch 'NIFI-4914'
Your branch is up to date with 'origin/NIFI-4914'.
Davids-MacBook-Pro:nifi david$ git rebase --continue
No rebase in progress?
`

MikeThomsen commented:
Pretty sure we found the culprit:

> Davids-MacBook-Pro:nifi david$ git rebase

That should be `git rebase master`. Not sure what git thinks it's doing, but it's clearly not doing what you want it to do. Give that a shot and let us know what you get.

david-streamlio commented:
Davids-MacBook-Pro:nifi david$ git checkout NIFI-4914
Switched to branch 'NIFI-4914'
Your branch is up to date with 'origin/NIFI-4914'.
Davids-MacBook-Pro:nifi david$ git rebase master
First, rewinding head to replay your work on top of it...
Applying: Added Apache Pulsar Processors and Controller Service
Applying: Changed code to use new ExpressionLanguageScope Enum
Applying: Changed artifact versions to 1.7.0-SNAPSHOT
Applying: Added Apache Pulsar Processors and Controller Service
Using index info to reconstruct a base tree...
M	nifi-nar-bundles/pom.xml
.git/rebase-apply/patch:848: trailing whitespace.
  
.git/rebase-apply/patch:854: trailing whitespace.
  
.git/rebase-apply/patch:857: trailing whitespace.
  
.git/rebase-apply/patch:859: space before tab in indent.
  	<dependency>
.git/rebase-apply/patch:865: trailing whitespace.
  
warning: squelched 161 whitespace errors
warning: 166 lines add whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/pom.xml
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/pom.xml
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/TestPublishPulsar_1_X.java
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/TestPublishPulsar_1_X.java
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/TestConsumePulsar_1_X.java
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/TestConsumePulsar_1_X.java
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/TestConsumePulsarRecord_1_X.java
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/TestConsumePulsarRecord_1_X.java
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/AbstractPulsarProcessorTest.java
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/AbstractPulsarProcessorTest.java
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsar_1_X.java
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsar_1_X.java
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsar_1_X.java
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsar_1_X.java
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarProducerProcessor.java
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarProducerProcessor.java
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarConsumerProcessor.java
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarConsumerProcessor.java
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/pom.xml
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/pom.xml
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-nar/pom.xml
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-nar/pom.xml
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/pom.xml
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/pom.xml
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service-nar/pom.xml
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service-nar/pom.xml
Auto-merging nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service-api/pom.xml
CONFLICT (add/add): Merge conflict in nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service-api/pom.xml
error: Failed to merge in the changes.
Patch failed at 0004 Added Apache Pulsar Processors and Controller Service
The copy of the patch that failed is found in: .git/rebase-apply/patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

MikeThomsen commented:
Bingo! That's what you should be seeing now. Fix the merge conflicts, `git add` the files and continue (and repeat until it's done)

david-streamlio commented:
Davids-MacBook-Pro:nifi david$ git rebase --continue
Applying: Changed artifact versions to 1.7.0-SNAPSHOT
Applying: Fixed issues identified during code review
Applying: Removed invalid characters left over from merge
Using index info to reconstruct a base tree...
M	nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/pom.xml
M	nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord_1_X.java
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
Davids-MacBook-Pro:nifi david$ git status
On branch NIFI-4914
Your branch and 'origin/NIFI-4914' have diverged,
and have 210 and 217 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

david-streamlio commented:
Well, that didn't go as well as we had hoped.

MikeThomsen commented:
Ok, I'll take a look. Worst comes to worst, we cherry-pick the heck out of your commits and play a game of hacky sack with Git and commits to get you back on track.

MikeThomsen commented:
Huh, I just checked out your latest branch and it's clean as a whistle WRT rebasing and appears 100% up to date w/ the upstream master. So not sure what the heck is going on here. Try doing a `git push origin --force NIFI-4914`. That should forcefully knock out whatever is going wrong.

MikeThomsen commented:
@david-streamlio Ok, I think I figured out what happened. At some point, it looks like you accidentally did a pull on upstream master into your branch. The fact that you keep having over 200 commits even with rebasing against master very strongly suggests that. What I did to verify was I checked out your branch, pushed it to my fork and saw 210ish commits with a merge conflict into my master. So I locally rebased against master, did a forced push and it dropped it down to ~7 commits.

So carefully follow the four steps I gave you:

1. git checkout master
2. git pull upstream master
3. git checkout NIFI-4914
4. git rebase master

**Make sure** that `upstream` is changed to whatever you call `apache/nifi` on github.com. The do:

`git push origin --force NIFI-4914` and it should all clear up to a few commits.

david-streamlio commented:
Followed all of the steps as you suggested. Performed a mvn -Pcontrib-check clean install without errors. 

Now we are failing CI due to missing content under the nifi-nar-bundles/nifi-grpc-bundle/nifi-grpc-processors/ folder.

joewitt commented:
@david-streamlio your PR appears to delete the nifi-grpc directory...You will want to avoid doing that.

MikeThomsen commented:
@joewitt I think if he were to rebase and drop [this commit](https://github.com/apache/nifi/pull/2614/commits/99a980cc1e6bbb45e64fb3431545023eefb8522c) it should fix that. What do you think?

joewitt commented:
I would rework this thing personally so that it gets back to just the commit(s) meant to be included.  Like start fresh off latest master, apply the actual changes intended, then end up with a single commit and either force push that OR send a diff PR.  This one is tough to follow as-is but seems like a good/cool/useful PR.

david-streamlio commented:
I tend to agree with @joewitt on this....Let's start from a fresh master and add in the new nifi-nar-bundle....It shouldn't be this hard to add in a completely fresh bundle, as the files are all net new and shouldn't have any upstream conflicts..

david-streamlio commented:
Abandoning this PR in favor of PR #2702

https://github.com/apache/nifi/pull/2702
